azurelinuxagent/common/protocol/wire.py

```diff
@@ -797,20 +797,26 @@ def update_goal_state(self, forced=False):
         except Exception as exception:
             raise ProtocolError("Error fetching goal state: {0}".format(ustr(exception)))
 
-    def update_extensions_goal_state(self):
+    def update_extensions_goal_state(self, wait_for_new_goal_state=True):
         try:
             # TODO: Remove this retry loop when invoking this method on each iteration of the goal state loop (currently it is invoked only on a new goal state)
-            for _ in range(3):
-                correlation_id = str(uuid.uuid4())
-                url, headers = self.get_host_plugin().get_vm_settings_request(correlation_id)
-                etag = self.get_etag()
-                if etag is not None:
-                    headers['if-none-match'] = etag
+            correlation_id = str(uuid.uuid4())
+            etag = self.get_etag()
 
+            for _ in range(3):
                 # TODO: Remove this log statement when invoking this method on each iteration of the goal state loop (currently it is invoked only on a new goal state)
                 logger.info("Fetching extensions goal state [correlation ID: {0} eTag: {1}]", correlation_id, etag)
+                def get_vm_settings():
+                    url, headers = self.get_host_plugin().get_vm_settings_request(correlation_id)
+                    if etag is not None:
+                        headers['if-none-match'] = etag
+                    return self.fetch(url, headers, ok_codes=[httpclient.OK, httpclient.NOT_MODIFIED])
 
-                vm_settings, response_headers = self.fetch(url, headers, ok_codes=[httpclient.OK, httpclient.NOT_MODIFIED])
+                try:
+                    vm_settings, response_headers = get_vm_settings()
+                except ResourceGoneError:
```

Comment:
This retry logic should  be in the HostPluginProtocol class, but currently we have different versions of it across our code base. I am still not sure of how much refactoring I will do of the networking code as part of fast track, but I will see if I can do some refactoring in the retry.